File: Data Analyst in R/Step 5 - Probability and Statistics/2. Statistics Intermediate in R Averages and Variability/5. Z-scores.r
## 1. Individual Values ##

library(readr)
# Heads up that the following instruction can generate a column specification warnings, you can ignore that.
# These warnings are related to that `readr` has misguidedly guessed the type of values in the column `Pool QC`. 
# To avoid this warning we can manually specify the type of this column the `col_types` parameter.
houses  <-  read_tsv('AmesHousing_1.txt', col_types = cols(`Pool QC` = col_character())) 

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 
    sqrt(sum(distances) / length(distances) )
}

st_dev  <-  standard_deviation(houses$SalePrice)
mean  <-  mean(houses$SalePrice)

library(ggplot2)
#ggplot(data = houses,
#    aes(x = SalePrice)) +
#    geom_density(alpha = 0.1, 
#                 color='blue', 
#                 fill='blue') +
#    geom_vline(aes(xintercept = mean, 
#                   color = 'Mean'), 
#               size = 1.2 ) +
#    geom_vline(aes(xintercept = 220000, 
#                   color = '220,000'), 
#               size = 1.2 ) +
#    geom_vline(aes(xintercept = mean+st_dev, 
#                   color = 'Standard deviation'), 
#               size = 1.2 ) +
#    scale_y_continuous(labels = scales::comma) +
#    scale_x_continuous(labels = scales::comma, 
#                       lim = c(min(houses$SalePrice), 
#                               max(houses$SalePrice))) +
#    scale_colour_manual(values = c("Mean"="black", 
#                                   "220,000"="red", 
#                                   "Standard deviation"="orange"), 
#                        name = "") +
#    theme_bw() + 
#    theme(legend.position='top') +
#    xlab("Sale Price") + 
#    ylab("Density")
very_expensive  <-  FALSE

## 2. Number of Standard Deviations ##

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 
    sqrt(sum(distances) / length(distances) )
}
distance  <-  220000 - mean(houses$SalePrice)
st_devs_away  <-  distance / standard_deviation(houses$SalePrice)

## 3. Z-scores ##

min_val  <-  min(houses$SalePrice)
mean_val  <-  mean(houses$SalePrice)
max_val  <-  max(houses$SalePrice) 

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 
    sqrt(sum(distances) / length(distances) )
}
z_score <- function(value, vector, bessel = FALSE) {
    mean  <-  mean(vector)
    
    st_dev <- ifelse(!bessel, sd(vector), standard_deviation(vector)) 
    
    (value - mean) / st_dev
}

min_z  <-  z_score(min_val, houses$SalePrice)
mean_z  <-  z_score(mean_val, houses$SalePrice)
max_z  <-  z_score(max_val, houses$SalePrice)

## 4. Locating Values in Different Distributions ##

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 
    sqrt(sum(distances) / length(distances) )
}

z_score <- function(value, vector, bessel = FALSE) {
    mean  <-  mean(vector)
    
    st_dev <- ifelse(!bessel, sd(vector), standard_deviation(vector)) 
    
    (value - mean) / st_dev
}

target_neighborhoods <- c('NAmes', 'CollgCr', 'OldTown', 'Edwards', 'Somerst')
# Filter only the interesting locations from the dataset
# Group by neighborhood and find the z-score for 200000 for every location
# Sort the result by zcore
library(dplyr)
houses %>%
    filter(Neighborhood %in% target_neighborhoods) %>%
    group_by(Neighborhood) %>%
    summarise(zscore = abs(z_score(200000, SalePrice))) %>%
    ungroup() %>%
    arrange(zscore)

# Find the location with the z-score closest to 0
best_investment  <-  'College Creek'

## 5. Transforming Distributions ##

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 
    sqrt(sum(distances) / length(distances) )
}

library(dplyr)
houses <- houses %>%
    mutate(z_prices = (SalePrice - mean(SalePrice)) / standard_deviation(SalePrice))
z_mean_price  <-  mean(houses$z_prices)
z_stdev_price  <-  standard_deviation(houses$z_prices)

# Transforming 'Lot Area'
houses <- houses %>%
    mutate(z_area = (`Lot Area` - mean(`Lot Area`)) / standard_deviation(`Lot Area`))

z_mean_area  <-  mean(houses$z_area)
z_stdev_area  <-  standard_deviation(houses$z_area)

## 6. The Standard Distribution ##

population  <-  c(0, 8, 0, 8)

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 
    sqrt(sum(distances) / length(distances) )
}
mean_pop  <-  mean(population)
stdev_pop  <-  standard_deviation(population)

standardized_pop  <-  (population - mean_pop) / stdev_pop

mean_z  <-  mean(standardized_pop)
stdev_z  <-  standard_deviation(standardized_pop)

## 7. Standardizing Samples ##

sample  <-  c(0, 8, 0, 8)

x_bar  <-  mean(sample)
s  <-  sd(sample)

standardized_sample  <-  (sample - x_bar) / s
stdev_sample  <-  sd(standardized_sample)

## 8. Using Standardization for Comparisons ##

standard_deviation <- function(vector) {
  distances  <-  (vector - mean(vector))**2 
  sqrt(sum(distances) / length(distances) )
}

houses <- houses %>%
  mutate(index_1 = SalePrice/ 100000 + 37) %>%
  mutate(index_1 = replace(index_1,  row_number() %% 2 != 0, NA_real_) ) %>%
  mutate(index_2 = SalePrice/ 90000 - 2.8) %>%
  mutate(index_2 = replace(index_2,  row_number() %% 2 == 0, NA_real_) )

mean_index_1  <-  mean(houses$index_1, na.rm = T)
mean_index_2  <-  mean(houses$index_2, na.rm = T)

sd_index_1  <-  standard_deviation(na.omit(houses$index_1))
sd_index_2  <-  standard_deviation(na.omit(houses$index_2))
houses <- houses %>%
  mutate(z_1 = (index_1 - mean_index_1) / sd_index_1 ) %>%
  mutate(z_2 = (index_2 - mean_index_2) / sd_index_2 ) 

head(houses %>% select(z_1, z_2), 2)

better  <-  'first'

## 9. Converting Back from Z-scores ##

houses_merged <- bind_rows(houses_1,houses_2) %>%
    mutate(z_1 = tidyr::replace_na(z_1,0)) %>%
    mutate(z_2 = tidyr::replace_na(z_2,0)) %>%
    mutate(z_merged = z_1 +  z_2)
mean  <-  50
st_dev  <-  10
houses_merged <- houses_merged %>%
    mutate(transformed = z_merged * st_dev + mean)
                            
mean_transformed  <-  mean(houses_merged$transformed)
stdev_transformed  <-  standard_deviation(houses_merged$transformed)

File: Data Scientist In Python/Step 5 - Probability and Statistics/2. Statistics Intermediate Averages and Variability/5. Z-scores.py
## 1. Individual Values ##

import pandas as pd
houses = pd.read_table('AmesHousing_1.txt')
import matplotlib.pyplot as plt
houses['SalePrice'].plot.kde(xlim = (houses['SalePrice'].min(),
                                    houses['SalePrice'].max()
                                    )
                            )

st_dev = houses['SalePrice'].std(ddof = 0)
mean = houses['SalePrice'].mean()
plt.axvline(mean, color = 'Black', label = 'Mean')
plt.axvline(mean + st_dev, color = 'Red', label = 'Standard deviation')
plt.axvline(220000, color = 'Orange', label = '220000')
plt.legend()

very_expensive = False

## 2. Number of Standard Deviations ##

distance = 220000 - houses['SalePrice'].mean()
st_devs_away = distance / houses['SalePrice'].std(ddof = 0)

## 3. Z-scores ##

min_val = houses['SalePrice'].min()
mean_val = houses['SalePrice'].mean()
max_val = houses['SalePrice'].max()
def z_score(value, array, bessel = 0):
    mean = sum(array) / len(array)
    
    from numpy import std
    st_dev = std(array, ddof = bessel)
    
    distance = value - mean
    z = distance / st_dev
    
    return z

min_z = z_score(min_val, houses['SalePrice'])
mean_z = z_score(mean_val, houses['SalePrice'])
max_z = z_score(max_val, houses['SalePrice'])

## 4. Locating Values in Different Distributions ##

def z_score(value, array, bessel = 0):
    mean = sum(array) / len(array)
    
    from numpy import std
    st_dev = std(array, ddof = bessel)
    
    distance = value - mean
    z = distance / st_dev
    
    return z
# Segment the data by location
north_ames = houses[houses['Neighborhood'] == 'NAmes']
clg_creek = houses[houses['Neighborhood'] == 'CollgCr']
old_town = houses[houses['Neighborhood'] == 'OldTown']
edwards = houses[houses['Neighborhood'] == 'Edwards']
somerset = houses[houses['Neighborhood'] == 'Somerst']

# Find the z-score for 200000 for every location
z_by_location = {}
for data, neighborhood in [(north_ames, 'NAmes'), (clg_creek, 'CollgCr'),
                     (old_town, 'OldTown'), (edwards, 'Edwards'),
                     (somerset, 'Somerst')]:
    
    z_by_location[neighborhood] = z_score(200000, data['SalePrice'],
                                          bessel = 0)

# Find the location with the z-score closest to 0
print(z_by_location)
best_investment = 'College Creek'

## 5. Transforming Distributions ##

mean = houses['SalePrice'].mean()
st_dev = houses['SalePrice'].std(ddof = 0)
houses['z_prices'] = houses['SalePrice'].apply(
    lambda x: ((x - mean) / st_dev)
    )
z_mean_price = houses['z_prices'].mean()
z_stdev_price = houses['z_prices'].std(ddof = 0)

# Transforming 'Lot Area'
mean_area = houses['Lot Area'].mean()
stdev_area = houses['Lot Area'].std(ddof = 0)
houses['z_area'] = houses['Lot Area'].apply(
    lambda x: ((x - mean_area) / stdev_area)
    )

z_mean_area = houses['z_area'].mean()
z_stdev_area = houses['z_area'].std(ddof = 0)

## 6. The Standard Distribution ##

from numpy import std, mean
population = [0,8,0,8]
mean_pop = mean(population)
stdev_pop = std(population, ddof = 0)

standardized_pop = []
for value in population:
    z = (value - mean_pop) / stdev_pop
    standardized_pop.append(z)
    
mean_z = mean(standardized_pop)
stdev_z = std(standardized_pop, ddof = 0)

## 7. Standardizing Samples ##

from numpy import std, mean
sample = [0,8,0,8]

x_bar = mean(sample)
s = std(sample, ddof = 1)

standardized_sample = []
for value in sample:
    z = (value - x_bar) / s
    standardized_sample.append(z)
stdev_sample = std(standardized_sample, ddof = 1)

## 8. Using Standardization for Comparisons ##

mean_index1 = houses['index_1'].mean()
stdev_index1 = houses['index_1'].std(ddof = 0)
houses['z_1'] = houses['index_1'].apply(lambda x: 
                                      (x - mean_index1) / stdev_index1
                                     )

mean_index2 = houses['index_2'].mean()
stdev_index2 = houses['index_2'].std(ddof = 0)
houses['z_2'] = houses['index_2'].apply(lambda x: 
                                      (x - mean_index2) / stdev_index2
                                     )

print(houses[['z_1', 'z_2']].head(2))
better = 'first'

## 9. Converting Back from Z-scores ##

mean = 50
st_dev = 10
houses['transformed'] = houses['z_merged'].apply(
                                lambda z: (z * st_dev + mean)
                                )
mean_transformed = houses['transformed'].mean()
stdev_transformed = houses['transformed'].std(ddof = 0)